src/persist/src/indexed/encoding.rs

```diff
@@ -0,0 +1,469 @@
+    /// Descriptions are half-open intervals `[lower, upper)`.
+    pub batches: Vec<(Description<u64>, String)>,
+}
+
+/// The structure serialized and stored as a value in [Blob] storage for
+/// data keys corresponding to future data.
+#[derive(Clone, Debug, Abomonation)]
+pub struct BlobFutureBatch {
+    /// Id of the stream this batch belongs to.
+    pub id: u64,
+    /// Which updates are included in this batch.
+    pub desc: Description<SeqNo>,
+    /// The updates themselves.
+    pub updates: Vec<(SeqNo, (String, String), u64, isize)>,
+}
+
+/// The structure serialized and stored as a value in [Blob] storage for data
+/// keys corresponding to trace data.
+///
+/// TODO: This probably wants to be a different level of abstraction, so we can
+/// put multiple small batches in a single blob but also break a very large
+/// batch over multiple blobs. We also may want to break the latter into chunks
+/// for checksum and encryption?
+#[derive(Clone, Debug, Abomonation)]
+pub struct BlobTraceBatch {
+    /// Id of the trace this batch belongs to.
+    pub id: u64,
+    /// Which updates are included in this batch.
+    pub desc: Description<u64>,
+    /// The updates themselves.
+    pub updates: Vec<((String, String), u64, isize)>,
+}
+
+impl Default for BlobMeta {
+    fn default() -> Self {
+        BlobMeta {
+            last_file_id: Self::now_millis(),
+            futures: Vec::new(),
+            traces: Vec::new(),
+        }
+    }
+}
+
+impl BlobMeta {
+    /// Asserts Self's documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        for (_, f) in self.futures.iter() {
+            f.validate()?;
+        }
+        for (_, t) in self.traces.iter() {
+            t.validate()?;
+        }
+        Ok(())
+    }
+
+    /// Returns the current SystemTime in millis since epoch.
+    pub fn now_millis() -> u128 {
+        SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap_or_default()
+            .as_millis()
+    }
+}
+
+impl BlobFutureMeta {
+    /// Asserts Self's documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        let mut prev: Option<&Description<SeqNo>> = None;
+        for (desc, _) in self.batches.iter() {
+            if let Some(prev) = prev {
+                if prev.upper() != desc.lower() {
+                    return Err(format!(
+                        "invalid batch sequence: {:?} followed by {:?}",
+                        prev, desc
+                    )
+                    .into());
+                }
+            }
+            prev = Some(desc)
+        }
+        Ok(())
+    }
+}
+
+impl BlobTraceMeta {
+    /// Asserts Self's documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
```

Comment:
this might be covered elsewhere (still reading) but i think we need to check that desc.upper >= desc.lower at each desc too?

Reply:
I'm 51/49 on this. Had it at first, and then decided to remove it as more Description's responsibility. If you're in favor though, that's enough to tip me back the other way.

Reply:
I was 50/50 on this (and even had it the other way at first). If you're in favor, good enough for me

Reply:
I'm not super attached to where we check that invariant, but I don't actually think its checked anywhere at the moment? I definitely think here is as good a place as any

Reply:
Oh also, do we want to allow empty intervals like [2, 2)? I think we should disallow it because then it makes the nonoverlapping-ness harder to reason about